### PR TITLE
Dropdown improvements

### PR DESCRIPTION
**Pull request by ArStah**
Hi, @mariolamacchia. 
I gonna do some improvements to your new dropdown component, here is the list:
- [x] Menu open/close animation
- [x] Close dropdown when re-clicking on head of dropdown, don't close when re-clicking on search input
- [x] Close dropdown when clicking on other dropdown
- [x] Don't close dropdown, when choosing option in multiple selection
- [x] Highlight the selected item in menu
- [x] Show `"No results found"` when there is no result in search
- [x] Add `maxSelections` prop for multiple selection dropdown
- [x] Add `icon` prop to dropdownItem
- [x] Show `flag`, `image` and `icon` in selected item
- [x] Show `flag`, `image` and `icon` in labels of selected items
- [x] Delete last value from selected when pressing `backspase` in search input of multiple selection dropdown
- [x] Open dropdown upward, if doesn't fit the viewport
- [x] Ability to force set direction
- [x] Add examples
- [x] Add tests
- [x] Use keyboard event to control dropdown

### CODE REVIEW

**Comment by ArStah** [comment]
I see there is a test failing on `should close the menu`, but it work correctly in browser, do you have any ideas why?

**Comment by ArStah** [comment]
Hmm @mariolamacchia, the bug with freezing the page when using a lot of options is only in your docs... In my project everything is working good

**Comment by ArStah** [comment]
I think for now it's ready to merge. Will open new PR when add more

File: docs/examples/modules/DropdownExample/DropdownDirection.example.vue [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  </sui-dropdown>
+</div>
+
+</template>
+
+<script>
+export default {
+  name: 'DropdownDirectionExample',
+};
+</script>
```
Comment (mariolamacchia) [inline docs/examples/modules/DropdownExample/DropdownDirection.example.vue]: Love this, but I think it should only change the default opening direction, while it should still go to the other side if there isn't enough space. Nobody wants the dropdown to go out of the screen.
The react implementation has a boolean `upward` prop, what do you think about that?

File: docs/examples/modules/DropdownExample/index.js [hunk 2]
```diff
@@ -19,12 +22,10 @@ export default [
...
-        title: 'Selection',
-        description: 'A dropdown can be used to select between choices in a form.',
-        component: Selection,
-      },
-      {
-        component: FriendSelection,
+        title: 'Dropdown opening direction',
+        info: 'By default, the dropdown menu opens downward, if the menu does not fit in the screen - it will open upward, but you can force set direction by specifying `direction` prop',
+        description: 'You can set the direction of opening dropdown menu',
+        component: DropdownDirection,
```
Comment (mariolamacchia) [inline docs/examples/modules/DropdownExample/index.js]: Ideally, the docs should contain the same examples of the original website. Can you try to keep the same structure? If you think we need some examples that are Vue-only related, they should go to a "Usage" section at the bottom.

File: docs/examples/modules/DropdownExample/SelectionWithIcons.example.vue [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        value: 1,
+      }, {
+        text: 'Female',
+        icon: 'female',
+        value: 2,
+      }],
+    };
+  },
+};
+</script>
```
Comment (mariolamacchia) [inline docs/examples/modules/DropdownExample/SelectionWithIcons.example.vue]: Could you use the same example of the original doc? https://semantic-ui.com/modules/dropdown.html#icon

File: src/modules/Dropdown/Dropdown.jsx [hunk 4]
```diff
@@ -32,10 +54,15 @@ export default {
...
+      type: Number,
+      default: Infinity,
+      description: 'Maximum possible selections when using multiple selection',
+    },
     options: {
       type: Array,
-      description: "Array of SuiDropdownItem props e.g. `{ text: '', value: '' }`",
-      required: true,
+      description: 'Array of SuiDropdownItem props e.g. `{ text: \'\', value: \'\' }`',
+      // required: true,
```
Comment (mariolamacchia) [inline src/modules/Dropdown/Dropdown.jsx]: I think this line can go

File: test/specs/modules/Dropdown/Dropdown.spec.js [hunk 5]
```diff
@@ -33,6 +37,61 @@ describe('Dropdown', () => {
...
+        options: [{ text: 'foo', value: 1 }],
+      },
+    });
+    wrapper.vm.setOpen();
+
+    wrapper.find('.text').trigger('click');
+
+    wrapper.vm.$nextTick(() => {
+      expect(wrapper.vm.open).to.equal(false);
+    });
```
Comment (mariolamacchia) [inline test/specs/modules/Dropdown/Dropdown.spec.js]: Are you sure you need to use `$nextTick`? Also note that you need to make the tests asynchronous when you add a callback/promise.

File: test/specs/modules/Dropdown/Dropdown.spec.js [hunk 6]
```diff
@@ -33,6 +37,61 @@ describe('Dropdown', () => {
     expect(openSpy).to.have.been.calledWith(false);
   });
 
+  it('should close the menu when clicking on option', () => {
+    const wrapper = shallow(DropdownWithRequired, {
+      propsData: {
+        options: [{ text: 'foo', value: 1 }],
+      },
+    });
+    wrapper.vm.setOpen(true);
```
Comment (mariolamacchia) [inline test/specs/modules/Dropdown/Dropdown.spec.js]: I wouldn't use a private method to test this. What about adding a function `openDropdown = (wrapper) => wrapper.trigger('click')`?

File: test/specs/modules/Dropdown/Dropdown.spec.js [hunk 7]
```diff
@@ -33,6 +37,61 @@ describe('Dropdown', () => {
...
+    const wrapper = shallow(DropdownWithRequired, {
+      propsData: {
+        multiple: true,
+        options: [{ text: 'foo', value: 1 }],
+      },
+    });
+
+    wrapper.vm.setOpen(true);
+    wrapper.find(DropdownItem).trigger('click');
+    expect(wrapper.vm.open).to.equal(true);
```
Comment (mariolamacchia) [inline test/specs/modules/Dropdown/Dropdown.spec.js]: Same as above, I'd try to see if the `open` class is there or something like that. We might want to change the internal behavior later on and this shouldn't need to change this unit test.

File: test/specs/modules/Dropdown/Dropdown.spec.js [hunk 8]
```diff
@@ -54,4 +113,217 @@ describe('Dropdown', () => {
...
+    const wrapper = shallow(DropdownWithRequired, {
+      propsData: {
+        multiple: true,
+        options: [{ text: 'foo', value: 1 }, { text: 'bar', value: 2 }],
+      },
+    });
+
+    wrapper.vm.$on('input', (value) => {
+      wrapper.setProps({ value });
+    });
```
Comment (mariolamacchia) [inline test/specs/modules/Dropdown/Dropdown.spec.js]: Do we need this?

File: test/specs/modules/Dropdown/Dropdown.spec.js [hunk 9]
```diff
@@ -54,4 +113,217 @@ describe('Dropdown', () => {
...
+    wrapper.find(DropdownItem).trigger('click');
+
+    const text = wrapper.find('div.text');
+
+    const icon = text.find(Icon);
+    expect(icon.is(Icon)).to.equal(true);
+    const flag = text.find(Flag);
+    expect(flag.is(Flag)).to.equal(true);
+    const image = text.find(Image);
+    expect(image.is(Image)).to.equal(true);
```
Comment (mariolamacchia) [inline test/specs/modules/Dropdown/Dropdown.spec.js]: For all the expectations like this, you can replace `image.is(Image)` with `image.exists()`

File: test/specs/modules/Dropdown/Dropdown.spec.js [hunk 10]
```diff
@@ -54,4 +113,217 @@ describe('Dropdown', () => {
...
+            image: {
+              src: '/test',
+            },
+          },
+        ],
+      },
+    });
+    wrapper.vm.$on('input', (value) => {
+      wrapper.setProps({ value });
+    });
```
Comment (mariolamacchia) [inline test/specs/modules/Dropdown/Dropdown.spec.js]: Not sure why you added all these event listeners

File: test/specs/modules/Dropdown/Dropdown.spec.js [hunk 11]
```diff
@@ -54,4 +113,217 @@ describe('Dropdown', () => {
...
+      },
+    });
+    const input = wrapper.find('input.search');
+    input.element.value = 'ba';
+    input.trigger('input');
+    expect(wrapper.vm.filteredOptions).to.deep.equal([{ text: 'bar' }, { text: 'baz' }]);
+    input.element.value = 'blah';
+    input.trigger('input');
+    expect(wrapper.vm.filteredOptions).to.deep.equal([]);
+    expect(wrapper.vm.message).to.equal('No results found');
```
Comment (mariolamacchia) [inline test/specs/modules/Dropdown/Dropdown.spec.js]: Same as above: `vm` is private and reflects an internal behavior. Instead, unit tests should only consider the output (ie: html, classes, events).

File: src/modules/Dropdown/Dropdown.jsx [hunk 12]
```diff
@@ -57,43 +84,88 @@ export default {
...
     };
   },
   computed: {
+    maximumValuesSelected() {
+      return this.multipleValue.length >= this.maxSelections;
+    },
+    downward() {
+      if (this.direction !== directions.auto) return this.direction === directions.downward;
+      this.calculateMenuDirection();
+      if (this.menuDirection === null) return true;
```
Comment (mariolamacchia) [inline src/modules/Dropdown/Dropdown.jsx]: I'd put brackets around here

File: src/modules/Dropdown/Dropdown.jsx [hunk 13]
```diff
@@ -32,10 +54,15 @@ export default {
       type: Boolean,
       description: 'A selection dropdown can allow multiple selections.',
     },
+    maxSelections: {
+      type: Number,
+      default: Infinity,
+      description: 'Maximum possible selections when using multiple selection',
+    },
```
Comment (mariolamacchia) [inline src/modules/Dropdown/Dropdown.jsx]: Not 100% sure about this. It isn't a functionality that Semantic UI provides, we shouldn't provide it either. Also, once you add one validation, you will be expected to add all kind of validations to your input fields.
We could consider a validation directive as a separate task. What are your thoughts?

**Comment by ArStah** [comment]
@mariolamacchia okay, i'll consider all of your notes

File: docs/examples/modules/DropdownExample/DropdownDirection.example.vue (hunk 1 above)
Comment (mariolamacchia) [reply docs/examples/modules/DropdownExample/DropdownDirection.example.vue]: I guess you're right: https://github.com/Semantic-Org/Semantic-UI/blob/master/src/definitions/modules/dropdown.js#L2407

At the moment they don't offer a way to do what you're proposing, though. I'd say let's go for `auto` and `upward`, like in the original API. I [opened an issue on SemanticUI](https://github.com/Semantic-Org/Semantic-UI/issues/6187) to see what they think about it: as soon as they point toward a possible solution we can implement it without waiting for their changes.

File: test/specs/modules/Dropdown/Dropdown.spec.js [hunk 14]
```diff
@@ -54,4 +111,231 @@ describe('Dropdown', () => {
...
+      },
+    });
+
+    wrapper.vm.$on('input', inputSpy);
+
+    const options = wrapper.findAll(DropdownItem);
+    options.at(0).trigger('click');
+    expect(inputSpy).to.have.been.calledWith(1);
+    options.at(1).trigger('click');
+    expect(inputSpy).to.have.been.calledWith(2);
```
Comment (mariolamacchia) [inline test/specs/modules/Dropdown/Dropdown.spec.js]: You can remove the `vm.$on` and write here:
```js
    options.at(0).trigger('click');
    options.at(1).trigger('click');
    expect(wrapper.emitted().input[0][0]).to.equal(1);
    expect(wrapper.emitted().input[1][0]).to.equal(2);
```

File: test/specs/modules/Dropdown/Dropdown.spec.js (hunk 8 above)
Comment (mariolamacchia) [reply test/specs/modules/Dropdown/Dropdown.spec.js]: You don't need to check that v-model works, as it's already well tested by vue itself. Want you want to know is if the `input` event is fired with the right value:

You can remove both the `vm.$on` and write (haven't tested myself):

```js
const options = wrapper.findAll(DropdownItem);
options.at(0).trigger('click');
expect(wrapper.emitted().input[0][0]).to.equal([1]);
wrapper.setProps({ value: [1] });
options.at(0).trigger('click');
expect(wrapper.emitted().input[0][1]).to.equal([1, 2]);
```

File: test/specs/modules/Dropdown/Dropdown.spec.js [hunk 15]
```diff
@@ -54,4 +111,231 @@ describe('Dropdown', () => {
...
+      propsData: {
+        multiple: true,
+        options: [{ text: 'foo', value: 1 }, { text: 'bar', value: 2 }],
+      },
+    });
+
+    wrapper.vm.$on('input', (value) => {
+      wrapper.setProps({ value });
+    });
+    wrapper.vm.$on('input', inputSpy);
```
Comment (mariolamacchia) [inline test/specs/modules/Dropdown/Dropdown.spec.js]: Same as above

File: test/specs/modules/Dropdown/Dropdown.spec.js [hunk 16]
```diff
@@ -54,4 +111,231 @@ describe('Dropdown', () => {
...
+        multiple: true,
+        maxSelections: 1,
+        options: [{ text: 'foo', value: 1 }, { text: 'bar', value: 2 }],
+      },
+    });
+
+    wrapper.vm.$on('input', (value) => {
+      wrapper.setProps({ value });
+    });
+    wrapper.vm.$on('input', inputSpy);
```
Comment (mariolamacchia) [inline test/specs/modules/Dropdown/Dropdown.spec.js]: Same as above

File: docs/examples/modules/DropdownExample/DropdownDirection.example.vue (hunk 1 above)
Comment (ArStah) [reply docs/examples/modules/DropdownExample/DropdownDirection.example.vue]: > At the moment they don't offer a way to do what you're proposing, though. 

Why? according to the [docs](https://semantic-ui.com/modules/dropdown.html#/settings) they have `auto`, `downward` and `upward` 

> When set to `auto` determines direction based on whether dropdown can fit on screen. Set to `upward` or `downward` to always force a direction.

i've checked it, when initialize dropdown with `direction: downward` - menu opens downward no matter what

File: src/modules/Dropdown/Dropdown.jsx (hunk 13 above)
Comment (mariolamacchia) [reply src/modules/Dropdown/Dropdown.jsx]: Sorry, I didn't see that

File: docs/examples/modules/DropdownExample/DropdownDirection.example.vue (hunk 1 above)
Comment (mariolamacchia) [reply docs/examples/modules/DropdownExample/DropdownDirection.example.vue]: Ok then, let's go for the `auto-upward` solution

**Comment by mariolamacchia** [comment]
Looks great now. Let's merge it as soon as the `auto-upward` is done 🙂

**Pull request merged**